### PR TITLE
Support wasm32 as compilation target

### PR DESCRIPTION
**Pull request by kwek20**
# Description of change

Allows compilation to `wasm32-unknown-unknown`

## Links to any relevant issues

fixes issue #1538 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

 tested on `wasm32-unknown-unknown` and  `wasm32-wasi` for
`--no-default-features --features="mnemonic, events, storage, message_interface, participation"`

### CODE REVIEW

**Comment by thibault-martinez** [comment]
Conflicts

File: Cargo.toml [hunk 1]
```diff
@@ -32,7 +32,8 @@ tokio = { version = "1.24.2", default-features = false, features = [ "sync", "rt
 zeroize = { version = "1.5.7", default-features = false, features = [ "zeroize_derive" ] }
 
 [target.'cfg(target_family = "wasm")'.dependencies]
-gloo-timers = { version = "0.2.6", default-features = false, features = [ "futures" ] }
+wasm-timer = { version = "0.2.5", default-features = false }
```
Comment (thibault-martinez) [inline Cargo.toml]: Are we sure about this? The author themself recommend against using it. What couldn't we do with `gloo-timers` that is supported by `rustwasm` ?

File: src/account/operations/helpers/mod.rs [hunk 2]
```diff
@@ -3,3 +3,6 @@
 
 /// The module for time related functions
 pub(crate) mod time;
+
```
Comment (thibault-martinez) [inline src/account/operations/helpers/mod.rs]:
```suggestion
```

File: src/account/operations/helpers/task.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2022 IOTA Stiftung
```
Comment (thibault-martinez) [inline src/account/operations/helpers/task.rs]:
```suggestion
// Copyright 2023 IOTA Stiftung
```

File: src/storage/adapter/wasm.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2021 IOTA Stiftung
```
Comment (thibault-martinez) [inline src/storage/adapter/wasm.rs]:
```suggestion
// Copyright 2022 IOTA Stiftung
```

File: Cargo.toml (hunk 1 above)
Comment (kwek20) [reply Cargo.toml]: HMm no, it was conflicting earlier but not anymore. ill revert 👍

Comment (kwek20) [reply Cargo.toml]: Yea gloo_timers was !Send, but thats no longer relevant 🥳

File: Cargo.toml [hunk 5]
```diff
@@ -33,6 +33,7 @@ zeroize = { version = "1.5.7", default-features = false, features = [ "zeroize_d
 
 [target.'cfg(target_family = "wasm")'.dependencies]
 gloo-timers = { version = "0.2.6", default-features = false, features = [ "futures" ] }
+gloo-storage = { version = "0.2.2", default-features = false }
```
Comment (thibault-martinez) [inline Cargo.toml:35]: Order!

File: src/account_manager/builder.rs [hunk 6]
```diff
@@ -19,11 +19,8 @@ use crate::events::EventEmitter;
 #[cfg(all(feature = "storage", not(feature = "rocksdb")))]
 use crate::storage::adapter::memory::Memory;
 #[cfg(feature = "storage")]
-use crate::{
-    account::handle::AccountHandle,
-    storage::{constants::default_storage_path, manager::ManagerStorage},
-};
-use crate::{account_manager::AccountManager, ClientOptions, Error};
+use crate::storage::{constants::default_storage_path, manager::ManagerStorage};
+use crate::{account::handle::AccountHandle, account_manager::AccountManager, ClientOptions, Error};
```
Comment (thibault-martinez) [inline src/account_manager/builder.rs]: `Error` is unused on wasm compilations

File: src/account/operations/helpers/task.rs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+#[cfg(target_family = "wasm")]
+pub(crate) fn spawn<F>(future: F) -> tokio::task::JoinHandle<F::Output>
+where
+    F: futures::Future + 'static,
+    F::Output: 'static,
+{
+    tokio::task::spawn_local(future)
+}
```
Comment (thibault-martinez) [inline src/account/operations/helpers/task.rs]: I think I would put this file closer to the root as it's also being used in account_manager

File: src/account/operations/transaction/high_level/sweep_outputs.rs [hunk 8]
```diff
@@ -100,95 +97,92 @@ impl AccountHandle {
...
+        //         ) {
+        //             outputs_to_sweep.push(output_data);
+        //         } else {
+        //     // do we care about them?
+        //             other_owned_outputs.push(output_data);
+        //         }
+        //     } else {
+        //         outputs_to_sweep.push(output_data);
+        //     }
+        // }
```
Comment (thibault-martinez) [inline src/account/operations/transaction/high_level/sweep_outputs.rs:143]: What's up with that?

File: src/storage/adapter/wasm.rs [hunk 9]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use super::StorageAdapter;
```
Comment (thibault-martinez) [inline src/storage/adapter/wasm.rs]: super after external

File: src/account/operations/transaction/high_level/sweep_outputs.rs (hunk 8 above)
Comment (kwek20) [reply src/account/operations/transaction/high_level/sweep_outputs.rs:143]: Not sure, its in our code: https://github.com/iotaledger/wallet.rs/blob/develop/src/account/operations/transaction/high_level/sweep_outputs.rs#L116

File: CHANGELOG.md [hunk 10]
```diff
@@ -31,8 +31,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `AccountSyncOptions, AliasSyncOptions, NftSyncOptions`;
 - `SyncOptions::{account, alias, nft}` fields;
 - `{TransactionOptions, TransactionOptionsDto}::burn`;
-- `Memory` storage adapter;
+- `Memory` and `Wasm` storage adapter;
```
Comment (thibault-martinez) [inline CHANGELOG.md]:
```suggestion
- `Memory` and `Wasm` storage adapters;
```

File: src/task.rs [hunk 11]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(not(target_family = "wasm"))]
+pub(crate) fn spawn<F>(future: F) -> tokio::task::JoinHandle<F::Output>
```
Comment (thibault-martinez) [inline src/task.rs:6]: I would maybe add a `#[inline(always)]` to both

File: src/lib.rs [hunk 12]
```diff
@@ -26,6 +26,9 @@ pub mod events;
 /// The storage module.
 pub mod storage;
```
Comment (thibault-martinez) [inline src/lib.rs]:
```suggestion
```

File: src/storage/adapter/wasm.rs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+impl WasmAdapter {
+    /// Initialises the storage adapter.
+    pub fn new() -> crate::Result<Self> {
+        Ok(Self(LocalStorage::new()))
+    }
+}
+
+#[async_trait::async_trait]
+impl StorageAdapter for WasmAdapter {
+    async fn get(&self, key: &str) -> crate::Result<String> {
```
Comment (thibault-martinez) [inline src/storage/adapter/wasm.rs:25]: Missing doc, also can you please copy paste docs from the other adapters, I slightly changed them.

File: src/storage/adapter/wasm.rs [hunk 14]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use gloo_storage::LocalStorage;
+
+use super::StorageAdapter;
+
+/// The storage id.
+pub const STORAGE_ID: &str = "Wasm";
```
Comment (Thoralf-M) [inline src/storage/adapter/wasm.rs:9]: Will this be the key in browsers local storage? If so, wouldn't it be better to call it something like `WasmWallet`?

Comment (kwek20) [reply src/storage/adapter/wasm.rs:9]: Right now its not used, couldnt verify how it looks yet in browsers. However its probably not needed at all. Ill remove it when i pr the full wasm bindings if thats the case

File: CHANGELOG.md [hunk 15]
```diff
@@ -54,6 +55,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `Response::{ParticipationEvents, ParticipationEvents}` contain `ParticipationEventWithNodes` instead of `(ParticipationEvent, Vec<Node>)`;
 - Remove `Error` suffix on some `Error` variants;
 - Exposed `FilterOptions` so it can be imported from `account::FilterOptions`;
+- Only expose `MessageHandler` methods `SetStrongholdPassword`, `SetStrongholdPasswordClearInterval` and `StoreMnemonic` when `feature = "stronghold` is enabled;
```
Comment (Thoralf-M) [inline CHANGELOG.md]:
```suggestion
- Only expose `MessageHandler` methods `SetStrongholdPassword`, `SetStrongholdPasswordClearInterval` and `StoreMnemonic` when `feature = "stronghold"` is enabled; 
```

**Review by Thoralf-M — DISMISSED** [review]

**Review by samuel-rufi — DISMISSED** [review]

File: src/account/operations/syncing/addresses/output_ids/mod.rs [hunk 16]
```diff
@@ -34,72 +35,114 @@ impl AccountHandle {
...
-                            .await
-                    })
-                    .await
-                }
-                .boxed(),
-            );
+            #[cfg(target_family = "wasm")]
+            {
+                results.push(
+                    self.clone()
```
Comment (thibault-martinez) [inline src/account/operations/syncing/addresses/output_ids/mod.rs]: That clone is a bit weird. `get_basic_output_ids_with_any_unlock_condition` shouldn't take ownership, can you change that ?

File: src/account/operations/syncing/addresses/output_ids/mod.rs [hunk 17]
```diff
@@ -34,72 +35,114 @@ impl AccountHandle {
...
```
Comment (thibault-martinez) [inline src/account/operations/syncing/addresses/output_ids/mod.rs]: same

File: src/account/operations/syncing/addresses/output_ids/mod.rs [hunk 18]
```diff
@@ -34,72 +35,114 @@ impl AccountHandle {
...
```
Comment (thibault-martinez) [inline src/account/operations/syncing/addresses/output_ids/mod.rs]: same

File: src/account/operations/syncing/addresses/output_ids/alias_foundry.rs [hunk 19]
```diff
@@ -12,65 +13,86 @@ use iota_client::{
...
+                let found_output_ids = res?;
+                output_ids.extend(found_output_ids);
+            }
         }
 
+        // Get all results
         if sync_options.alias.foundry_outputs {
-            let foundry_output_ids = self.get_foundry_output_ids(&output_ids).await?;
-            output_ids.extend(foundry_output_ids.into_iter());
+            let foundry_output_ids = &self.get_foundry_output_ids(&output_ids).await?;
```
Comment (thibault-martinez) [inline src/account/operations/syncing/addresses/output_ids/alias_foundry.rs]:
```suggestion
            let foundry_output_ids = self.get_foundry_output_ids(&output_ids).await?;
```

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**